dial.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2023 Sauce Labs Inc. All rights reserved.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package forwarder
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/netip"
+	"time"
+
+	"github.com/saucelabs/forwarder/log"
+)
+
+type DialConfig struct {
+	// DialTimeout is the maximum amount of time a dial will wait for
+	// connect to complete.
+	//
+	// With or without a timeout, the operating system may impose
+	// its own earlier timeout. For instance, TCP timeouts are
+	// often around 3 minutes.
+	DialTimeout time.Duration
+
+	// KeepAlive specifies the interval between keep-alive
+	// probes for an active network connection.
+	// If zero, keep-alive probes are sent with a default value
+	// (currently 15 seconds), if supported by the protocol and operating
+	// system. Network protocols or operating systems that do
+	// not support keep-alives ignore this field.
+	// If negative, keep-alive probes are disabled.
+	KeepAlive time.Duration
+
+	// DNS server(s) to use instead of system default.
+	// If provided with multiple servers, the first one is used as primary server, the rest are used as a fallback.
+	DNSServers []netip.AddrPort
+
+	// Timeout is the timeout for DNS queries. Works only with user-provided DNS servers.
+	// Each server is given ResolveTimeout to respond.
+	ResolveTimeout time.Duration
+}
+
+func DefaultDialConfig() DialConfig {
+	return DialConfig{
+		DialTimeout:    10 * time.Second,
+		KeepAlive:      30 * time.Second,
+		ResolveTimeout: 5 * time.Second,
+	}
+}
+
+func (c *DialConfig) Validate() error {
+	for i, u := range c.DNSServers {
+		if err := validateDNSAddress(u); err != nil {
+			return fmt.Errorf("servers[%d]: %w", i, err)
+		}
+	}
+	return nil
+}
+
+type dialFunc func(context.Context, string, string) (net.Conn, error)
+
+type Dialer struct {
+	cfg       DialConfig
+	nd        *net.Dialer
+	resolvers []net.Resolver
+	log       log.Logger
+}
+
+func NewDialer(cfg DialConfig, l log.Logger) (*Dialer, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	if len(cfg.DNSServers) > 0 {
+		l.Infof("using DNS servers: %s", cfg.DNSServers)
+	}
+
+	nd := &net.Dialer{
+		Timeout:   cfg.DialTimeout,
+		KeepAlive: cfg.KeepAlive,
+	}
+
+	resolvers := make([]net.Resolver, 0, len(cfg.DNSServers))
+	for _, s := range cfg.DNSServers {
+		dnsAddress := s.String()
+		resolvers = append(resolvers, net.Resolver{
+			PreferGo: true,
+			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+				l.Debugf("dial DNS server %s instead of %s", dnsAddress, address)
```

Comment:
What if I have 5 dns servers?
Would it fallback each of them?

Reply:
It would sequentially dial the 5 servers, one after another, unless a lookup succeeds. The net.Dialer dials two times by default. 
```
[proxy] DEBUG: dial DNS server 192.168.100.13:53 instead of 127.0.0.11:53
[proxy] DEBUG: dial DNS server 192.168.100.13:53 instead of 127.0.0.11:53
[proxy] DEBUG: failed to resolve wronghost on 192.168.100.13:53: lookup wronghost on 127.0.0.11:53: no such host
[proxy] ERROR: failed to round trip: lookup : failed to resolve wronghost on any DNS server
```